File: Pandacap.Data/RemoteActivityPubFavorite.fs
﻿namespace Pandacap.Data

open System
open System.ComponentModel.DataAnnotations

/// An image attachment to an ActivityPub post from a follow.
type RemoteActivityPubFavoriteImage() =
    member val Url = "" with get, set
    member val Name = nullString with get, set

/// A remote ActivityPub post that this app's instance owner has added to their Favorites.
type RemoteActivityPubFavorite() =
    [<Key>]
    member val LikeGuid = Guid.Empty with get, set

    member val ObjectId = "" with get, set
    member val CreatedBy = "" with get, set
    member val Username = nullString with get, set
    member val Usericon = nullString with get, set
    member val CreatedAt = DateTimeOffset.MinValue with get, set
    member val FavoritedAt = DateTimeOffset.MinValue with get, set
    member val Summary = nullString with get, set
    member val Sensitive = false with get, set
    member val Name = nullString with get, set
    member val Content = nullString with get, set
    member val Attachments = new ResizeArray<RemoteActivityPubFavoriteImage>() with get, set

    interface IPost with
        member this.Id = $"{this.LikeGuid}"
        member this.Usericon = this.Usericon
        member this.Username = this.Username
        member this.DisplayTitle =
            Option.ofObj this.Name
            |> Option.orElse (ExcerptGenerator.fromHtml this.Content)
            |> Option.defaultValue $"{this.ObjectId}"
        member this.Timestamp = this.CreatedAt
        member this.LinkUrl = this.ObjectId
        member this.ThumbnailUrls = this.Attachments |> Seq.map (fun a -> a.Url)

File: Pandacap.Data/RemoteActivityPubFavorite.cs
using System.ComponentModel.DataAnnotations;

namespace Pandacap.Data
{
    public class RemoteActivityPubFavorite : IPost
    {
        private static readonly Textify.HtmlToTextConverter _converter = new();

        [Key]
        public Guid LikeGuid { get; set; }

        [Required]
        public string ObjectId { get; set; } = "";

        [Required]
        public string CreatedBy { get; set; } = "";

        public string? Username { get; set; }
        public string? Usericon { get; set; }

        public DateTimeOffset CreatedAt { get; set; }
        public DateTimeOffset FavoritedAt { get; set; }

        public string? Summary { get; set; }
        public bool Sensitive { get; set; }

        public string? Name { get; set; }
        public string? Content { get; set; }

        public class ImageAttachment : IThumbnail, IThumbnailRendition
        {
            [Required]
            public string Url { get; set; } = "";

            public string? Name { get; set; }

            int IThumbnailRendition.Width => 0;
            int IThumbnailRendition.Height => 0;

            string? IThumbnail.AltText => Name;
            IEnumerable<IThumbnailRendition> IThumbnail.Renditions => [this];
        }

        public List<ImageAttachment> Attachments { get; set; } = [];

        string IPost.Id => $"{LikeGuid}";

        string? IPost.DisplayTitle
        {
            get
            {
                string? excerpt = Content == null
                    ? null
                    : _converter.Convert(Content)
                        .Split("\n", StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
                        .FirstOrDefault();

                if (excerpt != null && excerpt.Length > 60)
                    excerpt = excerpt[..60] + "...";

                return Name ?? excerpt ?? $"{LikeGuid}";
            }
        }

        string? IPost.LinkUrl => ObjectId;

        DateTimeOffset IPost.Timestamp => FavoritedAt;

        IEnumerable<IThumbnail> IPost.Thumbnails => Attachments;
    }
}
